sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -88,8 +88,11 @@ package object expressions  {
   /**
    * A helper function to bind given expressions to an input schema.
    */
-  def toBoundExprs(exprs: Seq[Expression], inputSchema: Seq[Attribute]): Seq[Expression] = {
-    exprs.map(BindReferences.bindReference(_, inputSchema))
+  def toBoundExprs[A <: Expression](
+    exprs: Seq[A],
+    inputSchema: Seq[Attribute]): Seq[A] = {
```

Comment:
why not just changing this to `AttributeSeq`?

Reply:
Sometimes this function is called with a zero-length exprs (```df.count```, for example). I am attempting to avoid constructing the AttributeSeq in that case, because AttributeSeq's constructor eagerly builds a data structure based on the attributes (```private val qualified3Part```).

Reply:
How expensive is it to build those structures on empty data? We could consider cachin an empty attribute seq and use that when the there are no attributes.

Reply:
I was thinking that just because exprs is empty, that would not necessarily mean inputSchema is empty. But experiments seem to indicate that inputSchema is also empty. So building the AttributeSeq would be extremely low-cost. I guess there is no reason to lazily build it.